lib/secretsscanner/scan/scan.go

```diff
@@ -0,0 +1,252 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package scan
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+
+	accessgraphsecretsv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1"
+	"github.com/gravitational/teleport/api/types/accessgraph"
+)
+
+// ScannerConfig specifies parameters for the scanner.
+type ScannerConfig struct {
+	Dirs []string
+	Log  *slog.Logger
+}
+
+// NewScanner creates a new scanner.
+func NewScanner(cfg ScannerConfig) (*Scanner, error) {
+	if len(cfg.Dirs) == 0 {
+		return nil, trace.BadParameter("missing dirs")
+	}
+	if cfg.Log == nil {
+		cfg.Log = slog.Default()
+	}
+	return &Scanner{
+		dirs:        cfg.Dirs,
+		log:         cfg.Log,
+		privateKeys: make(map[string]*accessgraphsecretsv1pb.PrivateKey),
+	}, nil
+}
+
+// Scanner is a scanner that scans directories for secrets.
+type Scanner struct {
+	dirs []string
+	log  *slog.Logger
+	// privateKeys is a map of private keys found during the scan.
+	// The key is the path to the private key file and the value is the private key representation.
+	privateKeys map[string]*accessgraphsecretsv1pb.PrivateKey
+}
+
+// ScanPrivateKeys scans directories for SSH private keys.
+func (s *Scanner) ScanPrivateKeys(ctx context.Context, deviceID string) []SSHPrivateKey {
```

Comment:
This method looks like it should be able to error. Instead of swallowing the error here, let the caller swallow it (if that's what they want).